Userland/xargs.cpp

```diff
@@ -0,0 +1,304 @@
+                    return Stop;
+                items_used_for_this_command = 0;
+                total_command_length = 0;
+                return Unget;
+            } else {
+                child_argv.append(strndup(item.characters_without_null_termination(), item.length()));
+                ++items_used_for_this_command;
+            }
+        }
+
+        return Continue;
+    });
+
+    if (!fail && !child_argv.is_empty())
+        fail = !run_command(child_argv, verbose, is_stdin, devnull_fd);
+
+    if (!is_stdin)
+        fclose(fp);
+
+    return fail ? 1 : 0;
+}
+
+bool read_items(FILE* fp, char entry_separator, Function<Decision(StringView)> callback)
+{
+    bool fail = false;
+
+    for (;;) {
+        char* item = nullptr;
+        size_t buffer_size = 0;
+
+        auto item_size = getdelim(&item, &buffer_size, entry_separator, fp);
+
+        if (item_size < 0) {
+            // getdelim() will return -1 and set errno to 0 on EOF.
+            if (errno != 0) {
+                perror("getdelim");
+                fail = true;
+            }
+            break;
+        }
+
+        Decision decision;
+        do {
+            decision = callback(item);
+            if (decision == Stop) {
+                free(item);
+                return true;
+            }
+        } while (decision == Unget);
+
+        free(item);
+    }
+
+    return fail;
+}
+
+bool run_command(Vector<char*>& child_argv, bool verbose, bool is_stdin, int devnull_fd)
+{
+    child_argv.append(nullptr);
+
+    if (verbose) {
+        StringBuilder builder;
+        builder.join(" ", child_argv);
+        fprintf(stderr, "xargs: %s\n", builder.to_string().characters());
+        fflush(stderr);
+    }
+
+    auto pid = fork();
+    if (pid < 0) {
+        perror("fork");
+        return false;
+    }
+
+    if (pid == 0) {
+        if (is_stdin)
+            dup2(devnull_fd, STDIN_FILENO);
+
+        execvp(child_argv[0], child_argv.data());
+        exit(1);
+    }
+
+    for (auto* ptr : child_argv)
+        free(ptr);
+
+    child_argv.clear();
+
+    int wstatus = 0;
+    if (waitpid(pid, &wstatus, 0) < 0) {
+        perror("waitpid");
+        return false;
+    }
+
+    if (WIFEXITED(wstatus)) {
+        if (WEXITSTATUS(wstatus) != 0)
+            return false;
+    } else {
+        return false;
+    }
+
+    return true;
+}
+
+ParsedInitialArguments::ParsedInitialArguments(Vector<const char*>& arguments, const StringView& placeholder)
+{
+    m_all_parts.ensure_capacity(arguments.size());
+    bool some_argument_has_replacement = false;
+
+    for (auto argument : arguments) {
+        StringView arg { argument };
+        Vector<StringView> parts;
+        if (placeholder.is_empty()) {
+            parts.append(arg);
+        } else {
+            auto maybe_placeholder_index = arg.find_first_of(placeholder);
+            size_t last_placeholder_end_offset = 0;
+            // Split the argument into segments delimited by the placeholder.
```

Comment:
Also hold ooon, why don't we just use `arg.split_view()`?

If that doesn't support strings, then well, it should; move this code to an `StringView::substring_view()` overload.

Reply:
It doesn't support strings, no.
I'm a little uncomfortable with messing with the stuff in AK, mainly due to the _ugh, rebuilding everything again_ factor of working on them

Reply:
You don't have to rebuild everything, `ninja xargs` should only rebuild what's necessary for you. (And it only takes a minute to rebuild the whole thing anyway — it took me longer to write this comment, and I'm not done typing it yet.)

The logic for splitting a string by a separator clearly belongs in AK. That's actually the same thing, separation of concerns — the splitting code shouldn't care about multiple arguments, and about adding an implicit placeholder at the end, and so on; and the placeholder parsing code shouldn't care about how exactly string splitting is implemented in terms of indexes and the like. But this time there's actually a chance of code reuse, though clearly this is not a common use case.

Reply:
By the way, you can clearly tell the simplification is working — I'm now able to understand what all this code is supposed to be doing, suggest improvements, and even recognize patterns like an ad-hoc `split_view()` implementation.

Reply:
Oh, I'm putting it in AK anyway, I'm just bitching about the annoyances of messing with AK